src/target/stm32f1_ch32.c

```diff
@@ -0,0 +1,338 @@
+{
+	DEBUG_CH("CH32: flash unlock \n");
+
+	target_mem_write32(t, FLASH_KEYR , KEY1);
+	target_mem_write32(t, FLASH_KEYR , KEY2);
+	// fast mode
+	target_mem_write32(t, FLASH_MODEKEYR_CH32 , KEY1);
+	target_mem_write32(t, FLASH_MODEKEYR_CH32 , KEY2);
+	uint32_t cr = target_mem_read32(t, FLASH_CR);
+	if (cr & FLASH_CR_FLOCK_CH32)
+	{
+		ERROR_CH("Fast unlock failed, cr: 0x%08" PRIx32 "\n", cr);
+		return -1;
+	}
+	return 0;
+}
+static int ch32f1_flash_lock(target *t)
+{
+	DEBUG_CH("CH32: flash lock \n");
+	SET_CR(FLASH_CR_LOCK);
+	return 0;
+}
+
+/**
+	\brief identify the ch32f1 chip
+	            Actually grab all cortex m3 with designer = arm not caught earlier...
+*/
+
+bool ch32f1_probe(target *t)
+{
+	t->idcode = target_mem_read32(t, DBGMCU_IDCODE) & 0xfff;
+	if ((t->cpuid & CPUID_PARTNO_MASK) != CORTEX_M3)
+		return false;
+	if(t->idcode !=0x410) { // only ch32f103
+		return false;
+	}
+
+	// try to flock
+	ch32f1_flash_lock(t);
+	// if this fails it is not a CH32 chip
+	if(ch32f1_flash_unlock(t)) {
+		return false;
+	}
+
+
+	uint32_t signature= target_mem_read32(t, FLASHSIZE);
+	uint32_t flashSize=signature & 0xFFFF;
+
+	target_add_ram(t, 0x20000000, 0x5000);
+	ch32f1_add_flash(t, FLASH_BEGIN_ADDRESS_CH32, flashSize*1024, 128);
+	target_add_commands(t, stm32f1_cmd_list, "STM32 LD/MD/VL-LD/VL-MD");
+	t->driver = "CH32F1 medium density (stm32f1 clone)";
+
+	// make sure we have 2 wait states
+	//target_mem_write32(t, FLASH_ACR,2);
+	return true;
+}
+/**
+  \fn ch32f1_flash_erase
+  \brief fast erase of CH32
+*/
+int ch32f1_flash_erase (struct target_flash *f,  target_addr addr, size_t len)
+{
+	target *t = f->t;
+	DEBUG_CH("CH32: flash erase \n");
+
+// Make sure we have 2 wait states, prefetch disabled
+	//target_mem_write32(t, FLASH_ACR , 2);
+
+	if (ch32f1_flash_unlock(t)) {
+		ERROR_CH("CH32: Unlock failed\n");
+		return -1;
+	}
+	// Fast Erase 128 bytes pages (ch32 mode)
+	while(len) {
+		SET_CR(FLASH_CR_FTER_CH32);// CH32 PAGE_ER
+		/* write address to FMA */
+		target_mem_write32(t, FLASH_AR , addr);
+		/* Flash page erase start instruction */
+		SET_CR( FLASH_CR_STRT );
+		WAIT_EOP();
+		CLEAR_EOP();
+		CLEAR_CR( FLASH_CR_STRT );
+		// Magic
+		MAGIC(addr);
+		if (len > 128)
+			len -= 128;
+		else
+			len = 0;
+		addr += 128;
+	}
+	sr = target_mem_read32(t, FLASH_SR);
+	ch32f1_flash_lock(t);
+	if ((sr & SR_ERROR_MASK)) {
+		ERROR_CH("ch32f1 flash erase error 0x%" PRIx32 "\n", sr);
+		return -1;
+	}
+	return 0;
+}
+
+/**
+	\fn waitFlashReady
+	\brief poll the beginning of a block till it is fffff, meaning we can proceeed
+
+*/
+
+static bool waitFlashReady(target *t,uint32_t adr)
+{
+  // Wait a bit
+  // the # of cycles needed to do N read access over SWD
+  // the actual number is around 10
+  // then check we do read ffff (/!\ even if we read ffff it does not mean its ok
+  // these are the data from the previous operation and they could be ffff)
+  uint32_t ff;
+  for(int i=0;i<32;i++)   ff=target_mem_read32(t,adr);
```

Comment:
Same comment as above about the lack of spaces here. Also, the body of the for loop really should be on its own line.